File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

typedef long long ll;
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return dfs(root, INT64_MIN, INT64_MAX);
    }
    
    bool dfs(TreeNode* root, ll left, ll right)
    {
        if(root == nullptr) return true;
        
        return root->val > left && root->val < right && dfs(root->left, left, root->val) && dfs(root->right, root->val, right);
    }
};

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        
        def valid(root, left, right):
            if not root:
                return True
            return root.val > left and root.val < right and valid(root.left, left, root.val) and valid(root.right, root.val, right)
        
        return valid(root, -10000000000, 10000000000)
        